src/main/java/ch/njol/skript/SkriptEventHandler.java

```diff
@@ -212,62 +213,111 @@ static ScriptInfo removeTriggers(File script) {
 				i--;
 			}
 		}
-		
+
 		info.commands = Commands.unregisterCommands(script);
-		
+
 		return info;
 	}
-	
+
 	static void removeAllTriggers() {
 		triggers.clear();
 		for (Trigger t : selfRegisteredTriggers)
 			((SelfRegisteringSkriptEvent) t.getEvent()).unregisterAll();
 		selfRegisteredTriggers.clear();
 	}
-	
+
 	/**
 	 * Registers event handlers for all events which currently loaded
 	 * triggers are using.
 	 */
-	@SuppressWarnings({"unchecked", "rawtypes"})
 	static void registerBukkitEvents() {
 		for (NonNullPair<Class<? extends Event>, Trigger> pair : triggers) {
 			assert pair.getFirst() != null;
 			Class<? extends Event> e = pair.getFirst();
-			
+
 			EventPriority priority;
 			priority = pair.getSecond().getEvent().getEventPriority();
-			
+
 			PriorityListener listener = listeners[priority.ordinal()];
 			EventExecutor executor = listener.executor;
-			
-			Set<Class<? extends Event>> registeredEvents = listener.registeredEvents;
-			
+
+			HandlerList handlerList = getHandlerList(e);
+
+			if (handlerList == null)
+				continue;
+
 			// PlayerInteractEntityEvent has a subclass we need for armor stands
 			if (e.equals(PlayerInteractEntityEvent.class)) {
-				if (!registeredEvents.contains(e)) {
-					registeredEvents.add(e);
+				if (!isEventRegistered(handlerList, priority)) {
 					Bukkit.getPluginManager().registerEvent(e, listener, priority, executor, Skript.getInstance());
 					Bukkit.getPluginManager().registerEvent(PlayerInteractAtEntityEvent.class, listener, priority, executor, Skript.getInstance());
 				}
 				continue;
 			}
-			if (e.equals(PlayerInteractAtEntityEvent.class) || e.equals(PlayerArmorStandManipulateEvent.class)) {
+			if (e.equals(PlayerInteractAtEntityEvent.class) || e.equals(PlayerArmorStandManipulateEvent.class))
 				continue; // Ignore, registered above
-			}
-			
-			if (!containsSuperclass((Set) registeredEvents, e)) { // I just love Java's generics
+
+			if (!isEventRegistered(handlerList, priority)) // Check if event is registered
 				Bukkit.getPluginManager().registerEvent(e, listener, priority, executor, Skript.getInstance());
-				registeredEvents.add(e);
+		}
+	}
+
+	/**
+	 * A cache for the getHandlerList methods of Event classes
+	 */
+	private static final Map<Class<? extends Event>, Method> handlerListMethods = Collections.synchronizedMap(new HashMap<>());
+
+	@Nullable
+	@SuppressWarnings("ThrowableNotThrown")
+	private static HandlerList getHandlerList(Class<? extends Event> eventClass) {
+		try {
+			Method method = getHandlerListMethod(eventClass);
+			method.setAccessible(true);
+			return (HandlerList) method.invoke(null);
+		} catch (Exception ex) {
+			Skript.exception(ex, "Failed to get HandlerList for event " + eventClass.getName());
+			return null;
+		}
+	}
+
+	private static Method getHandlerListMethod(Class<? extends Event> eventClass) {
+		Method method;
+
+		synchronized (handlerListMethods) {
```

Comment:
Does this need to be synchronized? I thought that was only for iterations

Reply:
It doesn't have to be, but not doing it may result in unnecessary work for the VM, consider for example 2 threads that both call getHandlerList method with the same class argument at roughly the same time, and this event's method isn't yet in cache. Below is a timeline:
- Thread 1 calls get on the cache map, receives null
- Thread 1 calls getHandlerListMethod_i
- Thread 2 calls on the get cache map, receives null
- Thread 2 calls getHandlerListMethod_i
- Thread 1 gets method from getHandlerListMethod_i and stores it in cache map
- Thread 2 gets method from getHandlerListMethod_i and stores it in cache map
While this behaviour won't break anything, IMO it's better not to make it get the method twice